libnavigation-core/src/test/java/com/mapbox/navigation/core/NavigationSessionTest.kt

```diff
@@ -1,156 +1,30 @@
-        navigationSession.onSessionStateChanged(TripSessionState.STARTED)
-        routes.add(route)
-        navigationSession.onRoutesChanged(routes)
-        navigationSession.onRoutesChanged(emptyList())
-        verify(exactly = 1) { accounts.navigationStopped() }
-    }
-
-    @Test
-    fun drive_route_noDrive() {
-        navigationSession.onSessionStateChanged(TripSessionState.STARTED)
-        routes.add(route)
-        navigationSession.onRoutesChanged(routes)
-        navigationSession.onSessionStateChanged(TripSessionState.STOPPED)
-        verify(exactly = 1) { accounts.navigationStopped() }
-    }
-
-    @Test
-    fun route_drive_noRoute() {
-        routes.add(route)
-        navigationSession.onRoutesChanged(routes)
-        navigationSession.onSessionStateChanged(TripSessionState.STARTED)
-        navigationSession.onRoutesChanged(emptyList())
-        verify(exactly = 1) { accounts.navigationStopped() }
-    }
-
-    @Test
-    fun route_drive_noDrive() {
-        routes.add(route)
-        navigationSession.onRoutesChanged(routes)
-        navigationSession.onSessionStateChanged(TripSessionState.STARTED)
-        navigationSession.onSessionStateChanged(TripSessionState.STOPPED)
-        verify(exactly = 1) { accounts.navigationStopped() }
-    }
-
-    @Test
-    fun route_drive_noRoute_noDrive() {
-        routes.add(route)
-        navigationSession.onRoutesChanged(routes)
-        navigationSession.onSessionStateChanged(TripSessionState.STARTED)
-        navigationSession.onRoutesChanged(emptyList())
-        navigationSession.onSessionStateChanged(TripSessionState.STOPPED)
-        verify(exactly = 1) { accounts.navigationStopped() }
-    }
-
-    @Test
-    fun route_drive_noDrive_noRoute() {
-        routes.add(route)
-        navigationSession.onRoutesChanged(routes)
-        navigationSession.onSessionStateChanged(TripSessionState.STARTED)
-        navigationSession.onSessionStateChanged(TripSessionState.STOPPED)
-        navigationSession.onRoutesChanged(emptyList())
-        verify(exactly = 1) { accounts.navigationStopped() }
-    }
-
-    @Test
-    fun restart_drive() {
-        routes.add(route)
-        navigationSession.onRoutesChanged(routes)
-        navigationSession.onSessionStateChanged(TripSessionState.STARTED)
-        navigationSession.onSessionStateChanged(TripSessionState.STOPPED)
-        navigationSession.onSessionStateChanged(TripSessionState.STARTED)
-        verify(exactly = 2) { accounts.navigationStarted() }
-    }
-
-    @Test
-    fun restart_route() {
-        routes.add(route)
-        navigationSession.onRoutesChanged(routes)
-        navigationSession.onSessionStateChanged(TripSessionState.STARTED)
-        navigationSession.onRoutesChanged(emptyList())
-        navigationSession.onRoutesChanged(routes)
-        verify(exactly = 2) { accounts.navigationStarted() }
-    }
-
     @Test
     fun stateObserverImmediateIdle() {
+        val navigationSession = NavigationSession()
```

Comment:
NIT: Why not prepare this in `@Before`? There's no preparation logic that would differ across tests.

Reply:
I'm not a big fan of inline setup in tests.

Reasons why 👇 

👀 [_Working Effectively with Unit Tests_](https://leanpub.com/wewut/read) 
> when writing tests you should prefer DAMP (Descriptive And Maintainable Procedures) 

That includes a bunch of _procedures_ as [Inline Setup](https://leanpub.com/wewut/read#leanpub-auto-inline-setup) 😅

**1. setUp() method**
- When you remove Setup you may limit the number of times you need to check variables when you first encounter a test. So then you no longer need to worry about irrelevant instances from other tests
- It's a waste of time because you normally need to understand all the Setup each time you have to deal with the test
- Setup advocates argue because if you don't use it you normally create duplication. You can solve it using private methods or builders

Normally is good to structure the tests following the triple A 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=40 http://wiki.c2.com/?ArrangeActAssert
If the `A`rrange part gets too long we should create factory or builder methods. Problem with Setup is that hides the _Arrange_ part completely.